File: ASP.NET/365ThreeSixty_v2/365ThreeSixty_v2/voteFactory.vb
Module voteFactory
    Public Function createVote(reviewerRef As Integer, recipientRef As Integer, comment As String)
        Dim reviewer As reviewer
        Dim recipient As recipient
        Dim v As vote

        v = New vote
        reviewer = createReviewer(reviewerRef)
        recipient = createRecipient(recipientRef)

        v.reviewer = reviewer.Ref
        v.recipient = recipient.Ref
        v.rawScore = 100
        v.tierFactor = getTierDeltaScore(reviewer.tier, recipient.tier)

        v.reviewerFactor = 0.9
        v.reviewerScore = getReviewerScore(v.reviewerFactor, reviewer.ReviewsInLast7Days)

        v.RecipientFactor = 0.9
        v.recipientScore = getRecipientScore(v.RecipientFactor, recipient.ReviewsInLast7Days)

        v.comment = comment
        v.commentScore = getCommentScore(comment)

        v.weightedScore = v.rawScore * v.tierFactor * v.reviewerScore * v.recipientScore * v.commentScore

        Return v
    End Function

    Public Function getTierDeltaScore(reviewerTier, recipientTier) As Double
        Dim DeltaScore As Double
        If reviewerTier >= recipientTier Then
            DeltaScore = 1 + (reviewerTier - recipientTier) / 10
        Else
            DeltaScore = 1 + (recipientTier - reviewerTier) / 10
        End If

        Return DeltaScore

    End Function

    Private Function getReviewerScore(factor As Double, ReviewsInLast7Days As Integer) As Double

        Dim ReviewerScore As Double

        ReviewerScore = factor ^ (ReviewsInLast7Days)
        Return ReviewerScore
    End Function

    Private Function getRecipientScore(factor As Double, ReviewsInLast7Days As Integer) As Double

        Dim recipientScore As Double

        recipientScore = factor ^ (ReviewsInLast7Days)
        Return recipientScore
    End Function

    Private Function getCommentScore(comment As String)



        '2 load review into array

        Dim reviewTable As New DataTable
        Dim reviewArray() As String
        comment = Replace(comment, ",", "")
        comment = Replace(comment, ".", "")
        comment = Replace(comment, ";", "")
        comment = Replace(comment, ":", "")


        reviewArray = comment.Split(" ")

        reviewTable.Columns.Add("commentWord")

        For Each word In reviewArray
            reviewTable.Rows.Add.Item("commentWord") = word
        Next


        'for each elemnent get a fuzzy percentage
        Dim k As Integer
        Dim r As Integer
        Dim matchRate As Single
        Dim totalMatchScore As Double

        totalMatchScore = 0
        For Each keyword In Form1.dsK.Tables(0).Rows
            For Each review In reviewTable.Rows

                matchRate = FuzzyPercent(keyword.item("keyword"), review.item("commentWord"))

                If matchRate >= 0.8 Then
                    totalMatchScore = totalMatchScore + 1
                End If

            Next

        Next
        Dim numberOfKeywords As Integer
        numberOfKeywords = Form1.dsK.Tables(0).Rows.Count

        If totalMatchScore > numberOfKeywords Then
            totalMatchScore = 2
        Else
            totalMatchScore = 1 + (totalMatchScore / numberOfKeywords)
        End If
        Return totalMatchScore
    End Function
End Module

File: 365ThreeSixtyAPI/365ThreeSixtyAPI/Factories/voteFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using _365ThreeSixtyAPI.Factories;
using _365ThreeSixtyAPI.Models;



namespace _365ThreeSixtyAPI.Factories

{
    public class voteFactory
    {
        public _365ThreeSixtyAPIContext db = new _365ThreeSixtyAPIContext();

        public voteResponse createVote(newVote newVote)
        {
            reviewerFactory rvFac = new reviewerFactory();
            recipientFactory rcFac = new recipientFactory();

            reviewer newReviewer = rvFac.createReviewer(newVote.reviewer, newVote.userAccountId);
            recipient newRecipient = rcFac.createRecipient(newVote.recipient, newVote.userAccountId);

            vote v = new vote();

            //set up
            v.userAccountId = newVote.userAccountId;
            v.reviewerRef = newReviewer.Ref;
            v.recipientRef = newRecipient.Ref;
            v.rawScore = 100;
            v.comment = newVote.comment;
            v.voteSubmittedAt = DateTime.Now;
            //factors, these will be used to adapt the weightings later on as part of the machine learning part##todo##
            v.tierFactor = 1;
            v.reviewerFactor = 1;
            v.recipientFactor = 1;
            v.commentFactor = 1;
            // derived scores
            v.tierScore = getTierScore(newReviewer.tier, newRecipient.tier);
            v.reviewerScore = getReviewerScore(newReviewer.reviewsInLastSevenDays);
            v.recipientScore = getRecipientScore(newRecipient.reviewsInLastSevenDays);
            v.commentScore = getCommentScore(newVote.comment);
            
            //final calculation
            v.weightedScore = v.rawScore * v.tierFactor * v.tierScore * v.reviewerFactor * v.reviewerScore * v.recipientFactor * v.recipientScore * v.commentFactor * v.commentScore;

            db.vote.Add(v);
            db.SaveChanges();

            voteResponse vr = new voteResponse();
            vr.recipient = newVote.recipient;
            vr.reviewer = newVote.reviewer;
            vr.voteMessage = "Your been recieved and counted, thanks";

            return vr;
        }

        private double getCommentScore(string comment)
        {
            //tidy and add words to a list
            char[] punc = { ',', '.', ';', ':', ' ', '?', '!' };
            string[] commentWords = comment.Split();

            List<string> commentList = new List<string>();

            foreach (string word in commentWords)
            {
                commentList.Add(word.TrimEnd(punc));
            }

            //get the keywords
            List<string> kw = db.keywordDictionary.Select(x => x.keyword).ToList();

            //compare the comment words with the keywords
            fuzzyMatch fm = new fuzzyMatch();
            double totalMatchScore = 0;
            foreach (string word in kw)
            {
                foreach (string com in commentWords)
                {
                    double matchrate = fm.FuzzyPercent(word, com);
                    if (matchrate > 0.8)
                    {
                        totalMatchScore = totalMatchScore + 1;
                    }

                }
            }

            int numberOfKeywords = kw.Count;
            if (totalMatchScore > numberOfKeywords)
            {
                totalMatchScore = 2;
            }
            else
            {
                totalMatchScore = 1 + (totalMatchScore / numberOfKeywords);
            }
               return totalMatchScore;
            }
        

        private double getRecipientScore(int reviewsInLastSevenDays)
        {
            double getRecipientScore = Math.Pow(0.9, reviewsInLastSevenDays);
            return getRecipientScore;
        }

        private double getReviewerScore(int reviewsInLastSevenDays)
        {
            double getReviewerScore = Math.Pow(0.9,reviewsInLastSevenDays);
            return getReviewerScore;
        }

        private double getTierScore(int reviewerTier, int recipientTier)
        {
            if (reviewerTier >= recipientTier)
            {
                double deltaScore = 1 + (reviewerTier - recipientTier) / 10;
                return deltaScore;
            }
            else
            {
                double deltaScore = 1 + (recipientTier - reviewerTier) / 10;
                return deltaScore;
            }

        }
    }
}